### PR TITLE
Fix execution context setting for iframes

### PR DESCRIPTION
**Pull request by ankur22**
## What?

This fixes a problem with setting the execution context when testing a website that uses iframes. Without this fix it could end up in an infinite loop when `waitForExecutionContext` is called.

## Why?

There is a race condition when it comes to attaching iframes and the execution context that apply to these frames. What usually occurs is:

1. An exec context for about:blank is first set;
2. A new set event is received for exec context for the url pointing to the actual destination for the iframe;
3. Finally the execution context for about:blank is destroyed, not for the second execution context.

This is the order of events when iframes are in use on a site, and so it is safe to nil the original execution context and overwrite it with the second one.

The exec context destroyed event will not remove the new exec context since the ids do not match.

If we didn't overwrite the first execCtx with the new one, then waitForExecutionContext could end up waiting indefinitely since all execCtx were destroyed.

Reference: [Playwright](https://github.com/microsoft/playwright/blob/352cb7fa08deaad87a594811f24306e509f83061/packages/playwright-core/src/server/frames.ts#L1612)

## Checklist

- [x] I have performed a self-review of my code
- [ ] I have added tests for my changes
- [x] I have commented on my code, particularly in hard-to-understand areas

## Related PR(s)/Issue(s)

Updates: https://github.com/grafana/xk6-browser/issues/1252

Since this fixes an interaction with `iframe`s, we now see more `panic: GoError: dispose: canceled` errors, which there is an open issue for: https://github.com/grafana/xk6-browser/issues/1089.

### CODE REVIEW

File: common/frame.go [hunk 1]
```diff
@@ -447,10 +447,31 @@ func (f *Frame) setContext(world executionWorld, execCtx frameExecutionContext)
...
+	// This is the order of events when iframes are in use on a site, and
+	// so it is safe to nil the original execution context and overwrite it
+	// with the second one.
+	//
+	// The exec context destroyed event will not remove the new exec context
+	// since the ids do not match.
+	//
+	// If we didn't overwrite the first execCtx with the new one, then
+	// waitForExecutionContext could end up waiting indefinitely since all
+	// execCtx were destroyed.
```
Comment (ankur22) [inline common/frame.go:468]: TODO: This needs an integration test

Comment (ankur22) [reply common/frame.go:468]: I was unable to create an integration test that replicates the issue that was in the original test script which can be found [here](https://community.grafana.com/t/getting-a-panic-error-in-test-after-clicking-place-order-button/115572). I'm having to skip the integration for this fix, but let me know if you have any ideas that I could try.

**Review by inancgumus — COMMENTED** [review]
Unfortunately, I still see the same error (after running the user's script with this branch) 😭 The reason you might not see the error is probably because of a race condition since the script uses a lot of `sleep`s.

**Comment by ankur22** [comment]
> Unfortunately, I still see the same error (after running the user's script with this branch) 😭 The reason you might not see the error is probably because of a race condition since the script uses a lot of sleeps.

Yep, you are correct, the same error that was posted in the original post still occurs 😢, when ran in headless mode:

```bash
panic: GoError: dispose: disposing element with ID -28511178728990825.29.1: Cannot find context with specified id (-32000)
```

However this same issue doesn't occur in headful mode with:

```bash
K6_BROWSER_HEADLESS=false k6 run e2e/test.js
```

Without this fix, in headful mode, the test ends up in an infinite loop when waitForExecutionContext is called. I think it's still a valid fix, but I will need to go back to the post and correct my assumption. I'll then look at resolving the original error 👍

**Comment by inancgumus** [comment]
> However this same issue doesn't occur in headful mode with:
> 
> ```shell
> K6_BROWSER_HEADLESS=false k6 run e2e/test.js
> ```
> 
> Without this fix, in headful mode, the test ends up in an infinite loop when waitForExecutionContext is called. I think it's still a valid fix, but I will need to go back to the post and correct my assumption. I'll then look at resolving the original error 👍

On the `main` branch, VUs get stuck after entering the email address. Yeah, this definitely fixes an issue 👏

On this fix's branch, when I run this fix in headful mode, I receive the following error. I also noticed that the Chromium processes didn't shut down.

```
panic: GoError: dispose: canceled

goroutine 403 [running]:
go.k6.io/k6/js/common.Throw(...)
        go.k6.io/k6@v0.50.0/js/common/util.go:20
github.com/grafana/xk6-browser/k6ext.Panic.func1(0x140028ee008, {0x14005e6cdb0?, 0x10220d601?, 0x14004670f50?})
        github.com/grafana/xk6-browser@v1.4.3/k6ext/panic.go:35 +0x7c
github.com/grafana/xk6-browser/k6ext.sharedPanic({0x102511488, 0x14006e26fa0}, 0x14006df5a40, {0x14005e6cdb0, 0x1, 0x1})
        github.com/grafana/xk6-browser@v1.4.3/k6ext/panic.go:64 +0x20c
github.com/grafana/xk6-browser/k6ext.Panic({0x102511488?, 0x14006e26fa0?}, {0x101d97046?, 0x102222420?}, {0x14005e6cdb0?, 0x102222420?, 0x140067e6660?})
        github.com/grafana/xk6-browser@v1.4.3/k6ext/panic.go:37 +0x58
github.com/grafana/xk6-browser/common.(*BaseJSHandle).Dispose(0x1400896d590)
        github.com/grafana/xk6-browser@v1.4.3/common/js_handle.go:83 +0xa0
github.com/grafana/xk6-browser/common.(*Frame).detach(0x1400e63a540)
        github.com/grafana/xk6-browser@v1.4.3/common/frame.go:243 +0x100
github.com/grafana/xk6-browser/common.(*FrameManager).removeFramesRecursively(0x14007104300, 0x1400e63a540)
        github.com/grafana/xk6-browser@v1.4.3/common/frame_manager.go:419 +0x1a8
github.com/grafana/xk6-browser/common.(*FrameManager).frameDetached(0x14007104300, {0x14004215aa0, 0x20}, {0x101d8eb21, 0x6})
        github.com/grafana/xk6-browser@v1.4.3/common/frame_manager.go:185 +0x110
github.com/grafana/xk6-browser/common.(*FrameSession).onFrameDetached(0x1400713c160, {0x14004215aa0, 0x20}, {0x101d8eb21, 0x6})
        github.com/grafana/xk6-browser@v1.4.3/common/frame_session.go:748 +0x11c
github.com/grafana/xk6-browser/common.(*FrameSession).initEvents.func1()
        github.com/grafana/xk6-browser@v1.4.3/common/frame_session.go:254 +0x664
created by github.com/grafana/xk6-browser/common.(*FrameSession).initEvents in goroutine 187
        github.com/grafana/xk6-browser@v1.4.3/common/frame_session.go:220 +0x180
```

What do you suggest about the next steps for issue #1252? I mean we can merge this PR, but keep the issue open. Or?

**Comment by ankur22** [comment]
> On this fix's branch, when I run this fix in headful mode, I receive the following error. I also noticed that the Chromium processes didn't shut down.
> 
> ```
> panic: GoError: dispose: canceled
> 
> goroutine 403 [running]:
> go.k6.io/k6/js/common.Throw(...)
>         go.k6.io/k6@v0.50.0/js/common/util.go:20
> github.com/grafana/xk6-browser/k6ext.Panic.func1(0x140028ee008, {0x14005e6cdb0?, 0x10220d601?, 0x14004670f50?})
>         github.com/grafana/xk6-browser@v1.4.3/k6ext/panic.go:35 +0x7c
> github.com/grafana/xk6-browser/k6ext.sharedPanic({0x102511488, 0x14006e26fa0}, 0x14006df5a40, {0x14005e6cdb0, 0x1, 0x1})
>         github.com/grafana/xk6-browser@v1.4.3/k6ext/panic.go:64 +0x20c
> github.com/grafana/xk6-browser/k6ext.Panic({0x102511488?, 0x14006e26fa0?}, {0x101d97046?, 0x102222420?}, {0x14005e6cdb0?, 0x102222420?, 0x140067e6660?})
>         github.com/grafana/xk6-browser@v1.4.3/k6ext/panic.go:37 +0x58
> github.com/grafana/xk6-browser/common.(*BaseJSHandle).Dispose(0x1400896d590)
>         github.com/grafana/xk6-browser@v1.4.3/common/js_handle.go:83 +0xa0
> github.com/grafana/xk6-browser/common.(*Frame).detach(0x1400e63a540)
>         github.com/grafana/xk6-browser@v1.4.3/common/frame.go:243 +0x100
> github.com/grafana/xk6-browser/common.(*FrameManager).removeFramesRecursively(0x14007104300, 0x1400e63a540)
>         github.com/grafana/xk6-browser@v1.4.3/common/frame_manager.go:419 +0x1a8
> github.com/grafana/xk6-browser/common.(*FrameManager).frameDetached(0x14007104300, {0x14004215aa0, 0x20}, {0x101d8eb21, 0x6})
>         github.com/grafana/xk6-browser@v1.4.3/common/frame_manager.go:185 +0x110
> github.com/grafana/xk6-browser/common.(*FrameSession).onFrameDetached(0x1400713c160, {0x14004215aa0, 0x20}, {0x101d8eb21, 0x6})
>         github.com/grafana/xk6-browser@v1.4.3/common/frame_session.go:748 +0x11c
> github.com/grafana/xk6-browser/common.(*FrameSession).initEvents.func1()
>         github.com/grafana/xk6-browser@v1.4.3/common/frame_session.go:254 +0x664
> created by github.com/grafana/xk6-browser/common.(*FrameSession).initEvents in goroutine 187
>         github.com/grafana/xk6-browser@v1.4.3/common/frame_session.go:220 +0x180
> ```

This shouldn't happen since the [fix](https://github.com/grafana/xk6-browser/pull/1261) was merged into `main` and `main` was rebased to this fix branch. Looking at the stack trace it seems that you may be running an older version of the branch?

```
> github.com/grafana/xk6-browser/common.(*BaseJSHandle).Dispose(0x1400896d590)
>         github.com/grafana/xk6-browser@v1.4.3/common/js_handle.go:83 +0xa0
```

In the latest version this points to a comment and not to a call to `panic`. Do you mind rebasing/pulling and trying again?

**Comment by inancgumus** [comment]
🤦 You were right. It somehow cached the version. I should have checked the versions in the stack trace. Apologies for the noise. WDYT about this?

> What do you suggest about the next steps for the issue https://github.com/grafana/xk6-browser/issues/1252? I mean, we can merge this PR but keep the issue open. Or?

**Review by inancgumus — APPROVED** [review]
Nice catch!

**Pull request merged**